### PR TITLE
Recaptcha enterprise general frontend protection

### PR DESCRIPTION
**Pull request by scarrero4660**
This fixes [oncokb/oncokb#3034](https://github.com/oncokb/oncokb/issues/3034)

### CODE REVIEW

File: src/main/resources/config/application-dev.yml
Comment (zhx828) [inline src/main/resources/config/application-dev.yml:180]: threshold and enabled need to have default value, otherwise they need to be String.
We could set to false and 0.

File: src/main/java/org/mskcc/cbio/oncokb/web/rest/CreateAssessment.java
Comment (zhx828) [inline src/main/java/org/mskcc/cbio/oncokb/web/rest/CreateAssessment.java:73]: What's the thought behind different case of logs?

**Review by zhx828 — CHANGES_REQUESTED** [review]
Why the registration and password reset endpoints are not protected by recaptcha? Also seems like both pages do not use recaptcha?

**Review by zhx828 — CHANGES_REQUESTED** [review]

File: src/main/webapp/app/App.tsx
Comment (scarrero4660) [reply src/main/webapp/app/App.tsx]: This is now deleted

File: src/main/java/org/mskcc/cbio/oncokb/config/application/FrontendProperties.java
Comment (scarrero4660) [reply src/main/java/org/mskcc/cbio/oncokb/config/application/FrontendProperties.java:13]: I removed this. If the recaptcha site key is not provided then a message will be logged and the token will not be validated. The page will now render even if the recaptcha properties are not set

Comment (zhx828) [reply src/main/java/org/mskcc/cbio/oncokb/config/application/FrontendProperties.java:13]: The github changes still show this prop

File: src/main/webapp/app/appConfig.ts
Comment (zhx828) [inline src/main/webapp/app/appConfig.ts:16]: The github changes still show this prop

File: src/main/webapp/app/shared/auth/RecaptchaBoundaryRoute.tsx
Comment (zhx828) [inline src/main/webapp/app/shared/auth/RecaptchaBoundaryRoute.tsx:41]: This is a class and this variable is not a state. You might run into issue that the view is not updated when the variable is.

File: src/main/webapp/app/shared/auth/RecaptchaBoundaryRoute.tsx
Comment (zhx828) [inline src/main/webapp/app/shared/auth/RecaptchaBoundaryRoute.tsx:13]: Can we clean up the imports?

**Review by zhx828 — CHANGES_REQUESTED** [review]

File: src/main/java/org/mskcc/cbio/oncokb/web/rest/RecaptchaResource.java
Comment (scarrero4660) [reply src/main/java/org/mskcc/cbio/oncokb/web/rest/RecaptchaResource.java]: It is not necessary. Now deleted

File: src/main/webapp/app/shared/auth/RecaptchaBoundaryRoute.tsx
Comment (scarrero4660) [reply src/main/webapp/app/shared/auth/RecaptchaBoundaryRoute.tsx]: Yes these are used to execute the recaptcha and store the validation value

**Comment by scarrero4660** [comment]
Starting next week we should implement a 0.5 score threshold for frontend and API endpoint protection. If the score is below the threshold, we should re-route to the homepage.

**Comment by zhx828** [comment]
@scarrero4660 some questions after testing the PR locally
- If you go into the pages that protected by recaptcha, you will encounter exception that is handled by a catch. For instance, http://localhost:9095/gene/IDH1/R132H/Acute%20Myeloid%20Leukemia I think it's because the recaptcha is not rendered before the loadData method is called
- how should we handle `recaptchaError`, cause it's not doing anything right now
- I don't think we need the `GOOGLE_APPLICATION_CREDENTIALS` any more? I was able to create assessment and validate without it. I'm a bit confused here
- Also is there a way to consolidate recaptcha properties under `ApplicationProperties` and `FrontendProperties`? I think we gonna use the same recaptcha key/project id for both frontend and backend.

**Comment by calvinlu3** [comment]
### Updates on maven tests not passing
I can verify that the maven tests fail if `GOOGLE_APPLICATION_CREDENTIALS` is not set. When running through the command line using `./mvnw test`, you need to set the env variable. If you are running via your IDE (using the green triangle next to test), then you need to configure your IDE to setup the env variables. For VSCode specifically, you can follow the instructions [HERE](https://code.visualstudio.com/docs/java/java-testing#_customize-test-configurations)

I tested using cmd line and through IDE and both work once the env variable is properly passed.


### Github Actions
We need to fetch the credentials file and use in Github action workflow somehow.
One option is to:
1. Create a github action secret with the  `GOOGLE_APPLICATION_CREDENTIALS` encoded in base64
2. Add a step to decode the base64 encoded string to a file. 
3. Export GOOGLE_APPLICATION_CREDENTIALS with value equal to the path
I can work on this to get the tests passing again.

@zhx828 I don't seem to have access to the settings of oncokb-public. I need to go into settings to add a new github actions secret.

**Comment by zhx828** [comment]
@calvinlu3 added you as the admin of oncokb-public. let me know whether that works.

**Comment by calvinlu3** [comment]
In order to run our `Application CI` check, we need to pass in our google credentials for recaptcha. However by default, secrets are not passed to workflows triggered from forks. This is to prevent malicious users from exfiltrating secrets. 

One way to overcome this restriction is to follow the guide [here](https://dev.to/petrsvihlik/using-environment-protection-rules-to-secure-secrets-when-building-external-forks-with-pullrequesttarget-hci)
Essentially, when a workflow is triggered via a pull request, an admin of this repo needs to approve it before the workflow can continue running. We can configure so that organization users can bypass this approval process, but collaborators from outside our organization have to get their PR approved to run the workflow. 
One potential issue with this is for outside collaborators. If they are approved to run the workflow, but later make a commit, then they should need approval again. This is to prevent them from making malicious changes after initial approval.

**Comment by calvinlu3** [comment]
### Task: Investigate why the reCAPTCHA token always return score 0.9

Here is the script used to test the `/register` endpoint and to scrape the genes page for alterations: [recaptcha.zip](https://github.com/oncokb/oncokb-public/files/12152305/recaptcha.zip)

I am testing with the recaptcha enterprise key named `oncokb-enterprise-dev-zhx` and the metrics show no `scores < 0.9`
image

The script also contains a method to scrape from google recaptcha demo website. However, I was unable to get a lower score. I tried to use a VPN and also used incognito mode for Selenium.

**Comment by zhx828** [comment]
ReCaptcha does not work very well for invisible. We plan to explore [hCAPTCHA](https://friendlycaptcha.com/insights/hcaptcha-vs-recaptcha/#:~:text=Many%20people%20are%20familiar%20with,solving%20them%20through%20machine%20learning.)

**Pull request closed without merging**
